File: frontend/lib/data/models/task.dart
import 'package:equatable/equatable.dart';
import 'user.dart';
import 'recurrence_rule.dart';

/// A household task.
class Task extends Equatable {
  final String id;
  final String householdId;
  final String title;
  final String? description;
  final List<User> assignedTo;
  final User? createdBy;
  final String status; // 'pending' | 'completed'
  final String priority; // 'low' | 'medium' | 'high'
  final String category; // cleaning | cooking | shopping | maintenance | other
  final DateTime? dueDate;

  /// Optional duration (PDR-004): both null = an instantaneous task
  /// (retrocompatible default). A recurring task never carries these — the
  /// backend ignores/clears them (duration + recurrence is out of scope this
  /// round), so the form hides the pickers whenever recurrence is on.
  final DateTime? startsAt;
  final DateTime? endsAt;

  final DateTime? completedAt;
  final User? completedBy;
  final bool isRecurring;
  final RecurrenceRule? recurrenceRule;
  final String? parentTaskId;

  /// Local-only: false while a create/update/delete made offline is still
  /// waiting in the pending-operations queue. Never sent to the server —
  /// see [toJson] — and irrelevant once the write has synced.
  final bool isSynced;

  /// True when this task is deleted — either the server's own soft-delete
  /// state (TD-009: the backend marks a task deleted instead of removing it,
  /// and the default task list excludes it, so this only ever comes back
  /// `true` from an explicit `includeDeleted=true` fetch, e.g. the trash
  /// view), or a local-only marker set while a delete made offline is still
  /// waiting in the pending-operations queue. Either way the row renders
  /// struck through — see [TaskTile] — until it is genuinely gone from the
  /// user's normal lists.
  final bool isDeleted;

  /// When the task was soft-deleted (TD-009), from the server. Null for an
  /// offline-only pending delete (the server hasn't recorded it yet) or a
  /// task that was never deleted.
  final DateTime? deletedAt;

  const Task({
    required this.id,
    required this.householdId,
    required this.title,
    this.description,
    this.assignedTo = const [],
    this.createdBy,
    this.status = 'pending',
    this.priority = 'medium',
    this.category = 'other',
    this.dueDate,
    this.startsAt,
    this.endsAt,
    this.completedAt,
    this.completedBy,
    this.isRecurring = false,
    this.recurrenceRule,
    this.parentTaskId,
    this.isSynced = true,
    this.isDeleted = false,
    this.deletedAt,
  });

  bool get isCompleted => status == 'completed';

  factory Task.fromJson(Map<String, dynamic> json) {
    return Task(
      id: (json['id'] ?? json['_id'] ?? '').toString(),
      householdId: (json['householdId'] ?? '').toString(),
      title: (json['title'] ?? '') as String,
      description: json['description'] as String?,
      assignedTo: (json['assignedTo'] as List<dynamic>?)
              ?.map(User.fromRef)
              .toList() ??
          const [],
      createdBy: json['createdBy'] != null ? User.fromRef(json['createdBy']) : null,
      status: (json['status'] ?? 'pending') as String,
      priority: (json['priority'] ?? 'medium') as String,
      category: (json['category'] ?? 'other') as String,
      dueDate:
          json['dueDate'] != null ? DateTime.tryParse(json['dueDate'].toString()) : null,
      startsAt:
          json['startsAt'] != null ? DateTime.tryParse(json['startsAt'].toString()) : null,
      endsAt: json['endsAt'] != null ? DateTime.tryParse(json['endsAt'].toString()) : null,
      completedAt: json['completedAt'] != null
          ? DateTime.tryParse(json['completedAt'].toString())
          : null,
      completedBy:
          json['completedBy'] != null ? User.fromRef(json['completedBy']) : null,
      isRecurring: (json['isRecurring'] ?? false) as bool,
      recurrenceRule: json['recurrenceRule'] != null
          // `Map.from`, not a cast. This map arrives from two places: an HTTP
          // response, where it really is a Map<String, dynamic>, and the Hive
          // cache, where NESTED maps come back as _Map<dynamic, dynamic> — the
          // adapter's `Map<String, dynamic>.from` only retypes the top level.
          // The hard cast that used to be here meant one cached recurring task
          // made `Hive.openBox<Task>` throw, which killed CacheService.init()
          // before runApp and left the app on a blank screen. Permanently: the
          // row stays on disk, so every relaunch failed the same way.
          ? RecurrenceRule.fromJson(
              Map<String, dynamic>.from(json['recurrenceRule'] as Map))
          : null,
      parentTaskId: json['parentTaskId']?.toString(),
      isSynced: (json['isSynced'] ?? true) as bool,
      isDeleted: (json['isDeleted'] ?? false) as bool,
      deletedAt:
          json['deletedAt'] != null ? DateTime.tryParse(json['deletedAt'].toString()) : null,
    );
  }

  /// Payload for create/update requests (only mutable fields). Deliberately
  /// excludes [isSynced]/[isDeleted]: they are local sync-queue bookkeeping,
  /// never something the server has an opinion on.
  Map<String, dynamic> toJson() => {
        'title': title,
        'description': description,
        'assignedTo': assignedTo.map((u) => u.id).toList(),
        'status': status,
        'priority': priority,
        'category': category,
        'dueDate': dueDate?.toIso8601String(),
        'startsAt': startsAt?.toIso8601String(),
        'endsAt': endsAt?.toIso8601String(),
        'isRecurring': isRecurring,
        if (recurrenceRule != null) 'recurrenceRule': recurrenceRule!.toJson(),
        'parentTaskId': parentTaskId,
      };

  Task copyWith({
    String? title,
    String? description,
    List<User>? assignedTo,
    String? status,
    String? priority,
    String? category,
    DateTime? dueDate,
    DateTime? startsAt,
    DateTime? endsAt,
    bool? isRecurring,
    RecurrenceRule? recurrenceRule,
    bool? isSynced,
    bool? isDeleted,
    DateTime? deletedAt,
  }) {
    return Task(
      id: id,
      householdId: householdId,
      title: title ?? this.title,
      description: description ?? this.description,
      assignedTo: assignedTo ?? this.assignedTo,
      createdBy: createdBy,
      status: status ?? this.status,
      priority: priority ?? this.priority,
      category: category ?? this.category,
      dueDate: dueDate ?? this.dueDate,
      startsAt: startsAt ?? this.startsAt,
      endsAt: endsAt ?? this.endsAt,
      completedAt: completedAt,
      completedBy: completedBy,
      isRecurring: isRecurring ?? this.isRecurring,
      recurrenceRule: recurrenceRule ?? this.recurrenceRule,
      parentTaskId: parentTaskId,
      isSynced: isSynced ?? this.isSynced,
      isDeleted: isDeleted ?? this.isDeleted,
      deletedAt: deletedAt ?? this.deletedAt,
    );
  }

  @override
  List<Object?> get props => [
        id,
        householdId,
        title,
        description,
        assignedTo,
        status,
        priority,
        category,
        dueDate,
        startsAt,
        endsAt,
        completedAt,
        isRecurring,
        parentTaskId,
        isSynced,
        isDeleted,
        deletedAt,
      ];
}

File: backend/src/models/Task.ts
import { Schema, model, Document, Types } from 'mongoose';
import { TaskStatus, TaskPriority, TaskCategory, RecurrenceType } from '../types';
import { jsonSchemaOptions } from '../utils/toJSON';

export interface IRecurrenceRule {
  type: RecurrenceType;
  interval?: number;
  daysOfWeek?: number[];
  dayOfMonth?: number;
}

/**
 * A household task. Indexed by householdId for fast per-household listing.
 * Supports assignment to multiple members and optional recurrence.
 */
export interface ITask extends Document {
  _id: Types.ObjectId;
  householdId: Types.ObjectId;
  title: string;
  description?: string;
  assignedTo: Types.ObjectId[];
  createdBy: Types.ObjectId;
  status: TaskStatus;
  priority: TaskPriority;
  category: TaskCategory;
  dueDate?: Date;
  /**
   * Optional duration (PDR-004): absent = an instantaneous task
   * (retrocompatible with every task created before this feature). A
   * recurring task never carries these — duration + recurrence is out of
   * scope this round, see task.service.ts.
   */
  startsAt?: Date;
  endsAt?: Date;
  completedAt?: Date;
  completedBy?: Types.ObjectId;
  isRecurring: boolean;
  recurrenceRule?: IRecurrenceRule;
  parentTaskId?: Types.ObjectId | null;
  /**
   * Soft delete (TD-009): DELETE marks a task deleted instead of removing the
   * document. Every read path excludes isDeleted:true by default — see
   * task.service.ts's listTasks/updateTask/completeTask — so a deleted task
   * behaves like a hard delete everywhere except the opt-in trash view
   * (`includeDeleted`) and the creator-or-admin restore endpoint.
   */
  isDeleted: boolean;
  deletedAt?: Date;
  createdAt: Date;
  updatedAt: Date;
}

const recurrenceRuleSchema = new Schema<IRecurrenceRule>(
  {
    type: { type: String, enum: ['daily', 'weekly', 'monthly', 'custom'] },
    interval: { type: Number, default: 1 },
    daysOfWeek: { type: [{ type: Number, min: 0, max: 6 }], default: undefined },
    dayOfMonth: { type: Number, min: 1, max: 31 },
  },
  { _id: false },
);

const taskSchema = new Schema<ITask>(
  {
    householdId: {
      type: Schema.Types.ObjectId,
      ref: 'Household',
      required: true,
      index: true,
    },
    title: { type: String, required: true, trim: true },
    description: { type: String },
    assignedTo: [{ type: Schema.Types.ObjectId, ref: 'User' }],
    createdBy: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    status: { type: String, enum: ['pending', 'completed'], default: 'pending', index: true },
    priority: { type: String, enum: ['low', 'medium', 'high'], default: 'medium' },
    category: {
      type: String,
      enum: ['cleaning', 'cooking', 'shopping', 'maintenance', 'other'],
      default: 'other',
    },
    dueDate: { type: Date },
    startsAt: { type: Date },
    endsAt: { type: Date },
    completedAt: { type: Date },
    completedBy: { type: Schema.Types.ObjectId, ref: 'User' },
    isRecurring: { type: Boolean, default: false },
    recurrenceRule: { type: recurrenceRuleSchema, default: undefined },
    // Links a generated occurrence back to the original recurring task.
    parentTaskId: {
      type: Schema.Types.ObjectId,
      ref: 'Task',
      required: false,
      default: null,
      index: true,
    },
    isDeleted: { type: Boolean, default: false, index: true },
    deletedAt: { type: Date },
  },
  { timestamps: true, ...jsonSchemaOptions },
);

// Matches the listing sort EXACTLY (householdId equality, then status desc,
// dueDate asc, _id desc). Directions must mirror the sort or MongoDB falls back
// to an in-memory sort, which is capped at 32MB and defeats pagination (TD-026).
// It also subsumes the previous { householdId, status, dueDate } index: every
// other query on those fields uses equality on status, where direction is
// irrelevant, so the old one was pure write overhead.
//
// isDeleted sits before dueDate, not after: it is an equality filter on
// EVERY query this index serves (listTasks excludes isDeleted:true by
// default, and the includeDeleted=true trash view still equality-filters on
// it, just for `true` instead) whereas dueDate is only ever a sort/range
// bound. Placing an always-equality field ahead of a range field keeps the
// index's bounds tight per MongoDB's Equality-Sort-Range guidance — a
// prefix that mixed the range field in before it would force a wider scan.
taskSchema.index({ householdId: 1, status: -1, isDeleted: 1, dueDate: 1, _id: -1 });

// Supports the PDR-003 timeline's from/to window query (GET /tasks?from=&to=),
// which filters by dueDate WITHOUT an equality predicate on status (the
// timeline shows pending and completed tasks together). The index above can't
// produce tight bounds for that shape: status sits between householdId and
// dueDate in its key, and a compound index only bounds a field tightly when
// every field before it is equality-matched (ESR rule) — leaving status
// unconstrained forces a bounds-less scan of the whole household on dueDate.
// This is a genuinely different key order, not a prefix of the index above,
// so it is not redundant with it (unlike the index TD-026 already removed).
//
// TD-064 extends it with `_id` so the keyset timeline's total order
// (`dueDate ASC, _id ASC`) is served entirely from the index: without the
// tiebreaker in the key, every page whose boundary falls inside a group of
// tasks sharing a dueDate would need an in-memory sort to resolve it, which
// is exactly where a cursor walk must not lose its footing.
//
// Note for whoever looks at Atlas: Mongoose only CREATES indexes, it never
// drops them, so the two-field `{householdId, dueDate}` version still exists
// there. It is a prefix of this one and therefore redundant — dropping it is
// a safe manual cleanup, not something this schema can do.
taskSchema.index({ householdId: 1, dueDate: 1, _id: 1 });

export const TaskModel = model<ITask>('Task', taskSchema);
